File: ts/project/.sdk/tm/rust/core/result.rs
// Operation result (mirrors go core/result.go).

use std::rc::Rc;

use crate::core::error::ProjectNameError;
use crate::core::helpers::{get_bool, get_str, getp, setp, to_int};
use crate::utility::voxgigstruct::Value;

/// Stream producer attached by the streaming feature: yields the result
/// items (or chunked batches) with configured pacing applied.
pub type StreamFn = Rc<dyn Fn() -> Vec<Value>>;

pub struct SdkResult {
    pub ok: bool,
    pub status: i64,
    pub status_text: String,
    pub headers: Value,
    pub body: Value,
    pub err: Option<ProjectNameError>,
    pub resdata: Value,
    pub resmatch: Value,

    // Feature extensions: pagination signals (paging feature) and the
    // incremental item iterator (streaming feature).
    pub paging: Value,
    pub streaming: bool,
    pub stream: Option<StreamFn>,
}

impl SdkResult {
    pub fn new(resmap: &Value) -> SdkResult {
        let ok = get_bool(resmap, "ok").unwrap_or(false);
        let status = match getp(resmap, "status") {
            Value::Noval => -1,
            s => to_int(&s),
        };
        let status_text = get_str(resmap, "statusText").unwrap_or_default();
        let headers = match getp(resmap, "headers") {
            Value::Map(m) => Value::Map(m),
            _ => Value::empty_map(),
        };
        let body = getp(resmap, "body");
        let resdata = getp(resmap, "resdata");
        let resmatch = match getp(resmap, "resmatch") {
            Value::Map(m) => Value::Map(m),
            _ => Value::Noval,
        };

        SdkResult {
            ok,
            status,
            status_text,
            headers,
            body,
            err: None,
            resdata,
            resmatch,
            paging: Value::Noval,
            streaming: false,
            stream: None,
        }
    }

    /// Match-map view for tests and explain records.
    pub fn to_value(&self) -> Value {
        let out = Value::empty_map();
        setp(&out, "ok", Value::Bool(self.ok));
        setp(&out, "status", Value::Num(self.status as f64));
        setp(&out, "statusText", Value::str(self.status_text.clone()));
        setp(&out, "headers", self.headers.clone());
        if !self.body.is_noval() {
            setp(&out, "body", self.body.clone());
        }
        if let Some(e) = &self.err {
            let em = Value::empty_map();
            setp(&em, "message", Value::str(e.msg.clone()));
            setp(&out, "err", em);
        }
        if !self.resdata.is_noval() {
            setp(&out, "resdata", self.resdata.clone());
        }
        if !self.resmatch.is_noval() {
            setp(&out, "resmatch", self.resmatch.clone());
        }
        if !self.paging.is_noval() {
            setp(&out, "paging", self.paging.clone());
        }
        out
    }
}

File: ts/project/.sdk/tm/zig/core/result.zig
// Operation result (mirrors go core/result.go).

const h = @import("helpers.zig");
const err = @import("error.zig");
const Value = h.Value;

// Stream producer attached by the streaming feature: yields result items
// (or chunked batches) with configured pacing applied.
pub const StreamFn = struct {
    ctx: *anyopaque,
    call: *const fn (ctx: *anyopaque) []Value,
};

pub const SdkResult = struct {
    ok: bool = false,
    status: i64 = -1,
    status_text: []const u8 = "",
    headers: Value,
    body: Value = .{ .null = {} },
    err: ?*err.ProjectNameError = null,
    resdata: Value = .{ .null = {} },
    resmatch: Value = .{ .null = {} },

    // Feature extensions.
    paging: Value = .{ .null = {} },
    streaming: bool = false,
    stream: ?StreamFn = null,

    pub fn make(resmap: Value) *SdkResult {
        const r = h.A().create(SdkResult) catch unreachable;
        const status: i64 = switch (h.getp(resmap, "status")) {
            .null => -1,
            else => |s| h.to_int(s),
        };
        const headers: Value = switch (h.getp(resmap, "headers")) {
            .object => h.getp(resmap, "headers"),
            else => h.omap(),
        };
        const resmatch: Value = switch (h.getp(resmap, "resmatch")) {
            .object => h.getp(resmap, "resmatch"),
            else => h.vnull(),
        };
        r.* = .{
            .ok = h.get_bool(resmap, "ok") orelse false,
            .status = status,
            .status_text = h.get_str(resmap, "statusText") orelse "",
            .headers = headers,
            .body = h.getp(resmap, "body"),
            .err = null,
            .resdata = h.getp(resmap, "resdata"),
            .resmatch = resmatch,
        };
        return r;
    }

    pub fn to_value(self: *const SdkResult) Value {
        const out = h.omap();
        h.setp(out, "ok", h.vbool(self.ok));
        h.setp(out, "status", h.vnum(self.status));
        h.setp(out, "statusText", h.vstr(self.status_text));
        h.setp(out, "headers", self.headers);
        if (!h.is_noval(self.body)) h.setp(out, "body", self.body);
        if (self.err) |e| {
            const em = h.omap();
            h.setp(em, "message", h.vstr(e.msg));
            h.setp(out, "err", em);
        }
        if (!h.is_noval(self.resdata)) h.setp(out, "resdata", self.resdata);
        if (!h.is_noval(self.resmatch)) h.setp(out, "resmatch", self.resmatch);
        if (!h.is_noval(self.paging)) h.setp(out, "paging", self.paging);
        return out;
    }
};
